### PR TITLE
Wip/hughsie/xb tool tokenize

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: src/xb-tool.c [hunk 1]
```diff
@@ -456,6 +481,8 @@ main (int argc, char *argv[])
 			"Return only when the silo is no longer valid", NULL },
 		{ "profile", 'p', 0, G_OPTION_ARG_NONE, &priv->profile,
 			"Show profiling information", NULL },
+		{ "tokenize", 'p', 0, G_OPTION_ARG_STRING_ARRAY, &priv->tokenize,
```
Comment (pwithnall) [inline src/xb-tool.c:484]: You can’t re-use the same `'p'` short name for the option here; it’s already used by `--profile`. Might also help to give an example of what values to provide to `--tokenize`.

File: src/xb-builder-node.c [hunk 2]
```diff
@@ -1219,3 +1222,26 @@ xb_builder_node_get_tokens (XbBuilderNode *self)
...
+	g_return_if_fail (tail_idx != XB_SILO_UNSET);
+
+	if (priv->token_idxs == NULL)
+		priv->token_idxs = g_array_new (FALSE, FALSE, sizeof(guint32));
+	g_array_append_val (priv->token_idxs, tail_idx);
+}
+
+/* private */
+GArray *
+xb_builder_node_get_token_idxs (XbBuilderNode *self)
```
Comment (pwithnall) [inline src/xb-builder-node.c:1242]: Might help to add a comment by this function saying that its return value is nullable.

File: src/xb-opcode-private.h [hunk 3]
```diff
@@ -12,14 +12,17 @@
 
 G_BEGIN_DECLS
 
+#define XB_OPCODE_TOKEN_MAX		32
```
Comment (pwithnall) [inline src/xb-opcode-private.h:15]: Could do with a comment mentioning what it’s for and why the value was chosen as 32.

File: src/xb-opcode.c [hunk 4]
```diff
@@ -257,6 +268,22 @@ xb_opcode_get_str (XbOpcode *self)
 	return self->ptr;
 }
 
+/**
+ * xb_opcode_get_tokens:
+ * @self: a #XbOpcode
+ *
+ * Gets the tokenized string stored on the opcode.
+ *
+ * Returns: a #GStrv, or %NULL if unset
```
Comment (pwithnall) [inline src/xb-opcode.c:277]: Add `(nullable)`

File: src/xb-string.c [hunk 5]
```diff
@@ -178,6 +178,32 @@ xb_string_search (const gchar *text, const gchar *search)
...
 
+/**
+ * xb_string_search: (skip)
+ * @text: The source string
+ * @search: The text to search for
+ *
+ * Searches for a fuzzy search match, ignoring search matches that are not at
+ * the start of the token.
+ *
+ * Returns: %TRUE if the string @search is contained in @text.
```
Comment (pwithnall) [inline src/xb-string.c:189]: Needs a `Since:` line

File: src/xb-string.c [hunk 6]
```diff
@@ -178,6 +178,32 @@ xb_string_search (const gchar *text, const gchar *search)
 	return FALSE;
 }
 
+/**
+ * xb_string_search: (skip)
+ * @text: The source string
+ * @search: The text to search for
+ *
+ * Searches for a fuzzy search match, ignoring search matches that are not at
+ * the start of the token.
```
Comment (pwithnall) [inline src/xb-string.c:187]: What’s a ‘fuzzy’ search match? Perhaps some examples would help.

File: src/xb-string.c [hunk 7]
```diff
@@ -178,6 +178,32 @@ xb_string_search (const gchar *text, const gchar *search)
 	return FALSE;
 }
 
+/**
+ * xb_string_search: (skip)
```
Comment (pwithnall) [inline src/xb-string.c:182]: `searchv` (the `v` is missing)

File: src/xb-string.c [hunk 8]
```diff
@@ -178,6 +178,32 @@ xb_string_search (const gchar *text, const gchar *search)
...
+ * @text: The source string
+ * @search: The text to search for
```
Comment (pwithnall) [inline src/xb-string.c:184]: This should probably say that `search` is actually a (`NULL` terminated?) array of somethings

File: src/xb-string.c [hunk 9]
```diff
@@ -178,6 +178,32 @@ xb_string_search (const gchar *text, const gchar *search)
 	return FALSE;
 }
 
+/**
+ * xb_string_search: (skip)
+ * @text: The source string
```
Comment (pwithnall) [inline src/xb-string.c:183]: This should probably say that `text` is actually a (`NULL` terminated?) array of somethings

File: src/xb-tool.c (hunk 1 above)
Comment (hughsie) [reply src/xb-tool.c:484]: Ahh, typo, fixed. I've also added an example, thanks.

File: src/xb-builder-node.c (hunk 2 above)
Comment (hughsie) [reply src/xb-builder-node.c:1242]: Fixed, ta.

File: src/xb-opcode-private.h (hunk 3 above)
Comment (hughsie) [reply src/xb-opcode-private.h:15]: Fixed.

File: src/xb-opcode.c (hunk 4 above)
Comment (hughsie) [reply src/xb-opcode.c:277]: I don't think it is -- it's a static array of size `XB_OPCODE_TOKEN_MAX+1` i.e. always allocated.

File: src/xb-string.c (hunk 5 above)
Comment (hughsie) [reply src/xb-string.c:189]: Fixed, thanks.

File: src/xb-string.c (hunk 6 above)
Comment (hughsie) [reply src/xb-string.c:187]: Added

File: src/xb-string.c (hunk 7 above)
Comment (hughsie) [reply src/xb-string.c:182]: Ta.

File: src/xb-string.c (hunk 8 above)
Comment (hughsie) [reply src/xb-string.c:184]: I've fixed up that and given an example.

File: src/xb-string.c (hunk 9 above)
Comment (hughsie) [reply src/xb-string.c:183]: Fixed.

File: src/xb-opcode.c (hunk 4 above)
Comment (pwithnall) [reply src/xb-opcode.c:277]: Ooops, I got mixed up with the `token_idxs` array, which is lazily allocated in `xb-builder-node.c`. Sorry!

Comment (pwithnall) [reply src/xb-opcode.c:277]: But the “or `%NULL` if unset” does need dropping in that case

File: src/xb-opcode.c [hunk 10]
```diff
@@ -413,6 +440,17 @@ xb_opcode_set_val (XbOpcode *self, guint32 val)
...
 }
 
+/* private */
+void
+xb_opcode_set_token (XbOpcode *self, guint idx, const gchar *val)
+{
+	g_return_if_fail (val != NULL);
+	g_return_if_fail (val[0] != '\0');
+	g_return_if_fail (idx < XB_OPCODE_TOKEN_MAX);
+	self->tokens[idx] = val;
```
Comment (pwithnall) [inline src/xb-opcode.c:450]: `self->tokens` is treated as a `NULL`-terminated array elsewhere in this file, so this function should ensure the `NULL` terminator is correct. That can’t happen if, for example, a caller calls `set_token(0, "foo")` then `set_token(15, "disjoint")`. It doesn’t happen now, but could happen later on if the code is refactored.

My suggestion would be to rename this `xb_opcode_append_token()` and drop the `idx` argument. Return a `gboolean` which is `TRUE` for success, or `FALSE` if the tokens array is full. That would simplify error handling in the callers (they would no longer have to check `XB_OPCODE_TOKEN_MAX` themselves) and might mean you could bring `XB_OPCODE_TOKEN_MAX` into this C file as a private implementation detail (although I haven’t worked through whether that’s possible).

File: src/xb-silo.c [hunk 11]
```diff
@@ -1077,6 +1077,47 @@ xb_silo_machine_fixup_attr_exists_cb (XbMachine *self,
 	return TRUE;
 }
 
+static gboolean
+xb_silo_machine_fixup_attr_search_tokn_cb (XbMachine *self,
```
[the remark this answers is not in the document]
Comment (hughsie) [reply src/xb-silo.c:1081]: Not really :) Fixed.

File: src/xb-opcode.c (hunk 4 above)
Comment (hughsie) [reply src/xb-opcode.c:277]: Agree, fixed.

File: src/xb-opcode.c (hunk 10 above)
Comment (hughsie) [reply src/xb-opcode.c:450]: > My suggestion would be to rename this xb_opcode_append_token() and drop the idx argument

Would we have to scan all the elements before adding each one? Doesn't that make each append check the NULLness a factorial number of times? Or do we make _XbOpcode slightly larger and store a `tokens_len` kind of thing?

Comment (pwithnall) [reply src/xb-opcode.c:450]: Probably add a `tokens_len`

Comment (hughsie) [reply src/xb-opcode.c]: Done!

**Comment by pwithnall** [comment]
LGTM.

**Pull request merged**